packages/ic-http-certification/src/http/http_request.rs

```diff
@@ -1,39 +1,766 @@
+    ///     .with_certificate_version(2)
+    ///     .build();
+    ///
+    /// assert_eq!(request.method(), "GET");
+    /// assert_eq!(request.url(), "/");
+    /// assert_eq!(request.headers(), &[("X-Custom-Foo".into(), "Bar".into())]);
+    /// assert_eq!(request.body(), &[1, 2, 3]);
+    /// assert_eq!(request.certificate_version(), Some(2));
+    /// ```
+    #[inline]
+    pub fn build(self) -> HttpRequest<'a> {
+        HttpRequest {
+            method: self.method.unwrap_or(Cow::Borrowed("GET")),
+            url: self.url.unwrap_or(Cow::Borrowed("/")),
+            headers: self.headers,
+            body: self.body,
+            certificate_version: self.certificate_version,
+        }
+    }
+
+    /// Build an [HttpUpdateRequest] from the builder.
+    ///
+    /// If the method is not set, it will default to `"GET"`.
+    /// If the URL is not set, it will default to `"/"`.
+    /// If the headers or body are not set, they will default to empty arrays.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ic_http_certification::HttpRequestBuilder;
+    ///
+    /// let update_request = HttpRequestBuilder::new()
+    ///     .with_method("GET")
+    ///     .with_url("/")
+    ///     .with_headers(vec![("X-Custom-Foo".into(), "Bar".into())])
+    ///     .with_body(&[1, 2, 3])
+    ///     .build_update();
+    ///
+    /// assert_eq!(update_request.method(), "GET");
+    /// assert_eq!(update_request.url(), "/");
+    /// assert_eq!(update_request.headers(), &[("X-Custom-Foo".into(), "Bar".into())]);
+    /// assert_eq!(update_request.body(), &[1, 2, 3]);
+    /// ```
+    #[inline]
+    pub fn build_update(self) -> HttpUpdateRequest<'a> {
+        HttpUpdateRequest {
+            method: self.method.unwrap_or(Cow::Borrowed("GET")),
+            url: self.url.unwrap_or(Cow::Borrowed("/")),
+            headers: self.headers,
+            body: self.body,
+        }
+    }
+}
+
+/// A Candid-encodable representation of an HTTP update request. This struct is
+/// used by the `http_update_request` method of the HTTP Gateway Protocol.
+///
+/// This is the same as [HttpRequest], excluding the
+/// [certificate_version](HttpRequest::certificate_version) property.
+///
+/// # Examples
+///
+/// ```
+/// use ic_http_certification::{HttpUpdateRequest, HttpRequest};
+///
+/// let request = HttpRequest::get("/")
+///     .with_method("GET")
+///     .with_url("/")
+///     .with_headers(vec![("X-Custom-Foo".into(), "Bar".into())])
+///     .with_body(&[1, 2, 3])
+///     .with_certificate_version(2)
+///     .build();
+/// let update_request = HttpUpdateRequest::from(request);
+///
+/// assert_eq!(update_request.method(), "GET");
+/// assert_eq!(update_request.url(), "/");
+/// assert_eq!(update_request.headers(), &[("X-Custom-Foo".into(), "Bar".into())]);
+/// assert_eq!(update_request.body(), &[1, 2, 3]);
+/// ```
+#[derive(Clone, Debug, CandidType, Deserialize, PartialEq, Eq)]
+pub struct HttpUpdateRequest<'a> {
```

Comment:
Not sure this is really needed, maybe let's discuss offline.

Reply:
It's not really needed in the canisters, but is used in the HTTP Gateway, currently from this crate: https://github.com/dfinity/agent-rs/blob/main/ic-utils/src/interfaces/http_request.rs#L57-L66.

I was thinking to eventually remove the HTTP stuff from that `ic-utils` crate and consolidate it into the `ic-http-certification` crate, or potentially break it out into a separate `ic-http` crate if there's a need for that.

Reply:
Yeah, let's chat offline 👍

Reply:
Chatted offline and agreed to keep the extra structure